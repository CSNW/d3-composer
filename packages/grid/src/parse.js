const PARSE_ROW = /\"(.*?)\"\s*([\w\d]+)/g;
const PARSE_MINMAX = /minmax\(([\w\d\s]+),([\w\d\s]+)\)/g;
const PARSE_REPEAT = /repeat\((\d+),([^\)]+)\)/g;

export function parseTemplate(spec) {
  const formatted = spec
    .split('\n')
    .map(trim)
    .filter(Boolean)
    .join(' ');
  const [areas_rows_spec, columns_spec] = formatted.split('/', 2).map(trim);

  if (!areas_rows_spec || !columns_spec) {
    throw new Error('Invalid grid template');
  }

  let areas_spec = '';
  let rows = [];
  let match;
  while ((match = PARSE_ROW.exec(areas_rows_spec))) {
    const [_, area, row] = match;

    areas_spec += `"${area}"`;
    rows.push(toLength(row));
  }

  let areas;
  if (!areas_spec) {
    rows = parseTracks(areas_rows_spec);
    areas = {};
  } else {
    areas = parseAreas(areas_spec);
  }

  const columns = parseTracks(columns_spec);

  return { areas, rows, columns };
}

export function parseTracks(spec) {
  let minmax_match;
  while ((minmax_match = PARSE_MINMAX.exec(spec))) {
    const [result, min, max] = minmax_match;

    // To handle potential nesting in repeat()
    // replace in spec with value that can be parsed by toLength
    const { index: start } = minmax_match;
    const end = start + result.length;
    const replacement = `MINMAX:${trim(min)}-${trim(max)}`;
    spec = spec.substr(0, start) + replacement + spec.substr(end);
  }

  let repeat_match;
  while ((repeat_match = PARSE_REPEAT.exec(spec))) {
    const [result, count, value] = repeat_match;
    const { index: start } = repeat_match;
    const end = start + result.length;

    const replacement = Array(parseInt(count, 10) + 1)
      .join(`${value} `)
      .trim();

    spec = spec.substr(0, start) + replacement + spec.substr(end);
  }

  const tracks = spec
    .split(' ')
    .filter(Boolean)
    .map(toLength);

  return tracks;
}

export function parseAreas(spec) {
  const areas = {};

  spec
    .split('"')
    .filter((_, i) => i % 2 !== 0)
    .forEach((row_spec, row, rows) => {
      const columns = row_spec.split(' ').filter(Boolean);

      columns.forEach((name, column) => {
        if (name === '.') return;

        const next_row = rows[row + 1];
        const next_column = columns[column + 1];

        if (!areas[name]) {
          areas[name] = { top: row };
        }
        if (!('left' in areas[name])) {
          areas[name].left = column;
        }
        if (!next_row || next_row[column] !== name) {
          areas[name].bottom = row;
        }
        if (!next_column || next_column !== name) {
          areas[name].right = column;
        }
      });
    });

  return areas;
}

export function toLength(value) {
  // units = 'fr' | 'percentage' | 'scalar'

  if (value.indexOf('MINMAX') === 0) {
    const [_, range] = value.split(':');
    const [min, max] = range.split('-');

    return [toLength(min), toLength(max)];
  }
  if (value === 'auto') {
    // For css grid, fr + auto would shrink auto to content
    // doesn't work well for this case
    // -> treat auto = 1fr
    return { value: 1, units: 'fr' };
  }
  if (endsWith(value, 'fr')) {
    return { value: Number(value.slice(0, -2)), units: 'fr' };
  }
  if (endsWith(value, '%')) {
    return { value: Number(value.slice(0, -1)) / 100, units: 'percentage' };
  }

  try {
    return { value: Number(value), units: 'scalar' };
  } catch (err) {
    throw new Error(`Unknown length value "${value}"`);
  }
}

function trim(value) {
  return value.trim();
}

function endsWith(value, check) {
  if (!value || !check) return false;

  const index = value.indexOf(check);
  if (index < 0) return false;

  return index === value.length - check.length;
}
