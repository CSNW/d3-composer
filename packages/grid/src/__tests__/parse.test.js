import { parseTracks, parseAreas, parseTemplate } from '../parse';

test('should parse tracks', () => {
  expect(parseTracks('1fr 1fr 1fr')).toMatchSnapshot();
  expect(parseTracks('1fr auto 20 30.5%')).toMatchSnapshot();
  expect(parseTracks(' 1  2%  3fr  auto ')).toMatchSnapshot();
});

test('should parse minmax in tracks', () => {
  expect(parseTracks(`1fr minmax(   20,1fr) 1fr`)).toMatchSnapshot();
});

test('should parse repeat in tracks', () => {
  expect(parseTracks(`1fr repeat(3, 0.5fr) 1fr`)).toMatchSnapshot();
});

test('should parse repeat and minmax in tracks', () => {
  expect(parseTracks(`repeat(4, minmax(20,1fr)    )`)).toMatchSnapshot();
});

test('should parse areas', () => {
  expect(
    parseAreas(`
      " a a a "
      " b c d "
      " b . e "
    `)
  ).toMatchSnapshot();
  expect(parseAreas('"a a a""b c d"  "b . e"')).toMatchSnapshot();
});

test('should parse template', () => {
  expect(
    parseTemplate(`
      " a a a " 50
      " b c d"auto
      " b . e " 50/50 auto 50
    `)
  ).toMatchSnapshot();

  expect(parseTemplate('50 auto 50/25 1fr 25%')).toMatchSnapshot();
});
