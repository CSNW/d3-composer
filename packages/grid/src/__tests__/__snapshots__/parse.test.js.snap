// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse areas 1`] = `
Object {
  "a": Object {
    "bottom": 0,
    "left": 0,
    "right": 2,
    "top": 0,
  },
  "b": Object {
    "bottom": 2,
    "left": 0,
    "right": 0,
    "top": 1,
  },
  "c": Object {
    "bottom": 1,
    "left": 1,
    "right": 1,
    "top": 1,
  },
  "d": Object {
    "bottom": 1,
    "left": 2,
    "right": 2,
    "top": 1,
  },
  "e": Object {
    "bottom": 2,
    "left": 2,
    "right": 2,
    "top": 2,
  },
}
`;

exports[`should parse areas 2`] = `
Object {
  "a": Object {
    "bottom": 0,
    "left": 0,
    "right": 2,
    "top": 0,
  },
  "b": Object {
    "bottom": 2,
    "left": 0,
    "right": 0,
    "top": 1,
  },
  "c": Object {
    "bottom": 1,
    "left": 1,
    "right": 1,
    "top": 1,
  },
  "d": Object {
    "bottom": 1,
    "left": 2,
    "right": 2,
    "top": 1,
  },
  "e": Object {
    "bottom": 2,
    "left": 2,
    "right": 2,
    "top": 2,
  },
}
`;

exports[`should parse minmax in tracks 1`] = `
Array [
  Object {
    "units": "fr",
    "value": 1,
  },
  Object {
    "units": "MINMAX",
    "values": Array [
      Object {
        "units": "scalar",
        "value": 20,
      },
      Object {
        "units": "fr",
        "value": 1,
      },
    ],
  },
  Object {
    "units": "fr",
    "value": 1,
  },
]
`;

exports[`should parse template 1`] = `
Object {
  "areas": Object {
    "a": Object {
      "bottom": 0,
      "left": 0,
      "right": 2,
      "top": 0,
    },
    "b": Object {
      "bottom": 2,
      "left": 0,
      "right": 0,
      "top": 1,
    },
    "c": Object {
      "bottom": 1,
      "left": 1,
      "right": 1,
      "top": 1,
    },
    "d": Object {
      "bottom": 1,
      "left": 2,
      "right": 2,
      "top": 1,
    },
    "e": Object {
      "bottom": 2,
      "left": 2,
      "right": 2,
      "top": 2,
    },
  },
  "columns": Array [
    Object {
      "units": "scalar",
      "value": 50,
    },
    Object {
      "units": "fr",
      "value": 1,
    },
    Object {
      "units": "scalar",
      "value": 50,
    },
  ],
  "rows": Array [
    Object {
      "units": "scalar",
      "value": 50,
    },
    Object {
      "units": "fr",
      "value": 1,
    },
    Object {
      "units": "scalar",
      "value": 50,
    },
  ],
}
`;

exports[`should parse template 2`] = `
Object {
  "areas": Object {},
  "columns": Array [
    Object {
      "units": "scalar",
      "value": 25,
    },
    Object {
      "units": "fr",
      "value": 1,
    },
    Object {
      "units": "percentage",
      "value": 0.25,
    },
  ],
  "rows": Array [
    Object {
      "units": "scalar",
      "value": 50,
    },
    Object {
      "units": "fr",
      "value": 1,
    },
    Object {
      "units": "scalar",
      "value": 50,
    },
  ],
}
`;

exports[`should parse tracks 1`] = `
Array [
  Object {
    "units": "fr",
    "value": 1,
  },
  Object {
    "units": "fr",
    "value": 1,
  },
  Object {
    "units": "fr",
    "value": 1,
  },
]
`;

exports[`should parse tracks 2`] = `
Array [
  Object {
    "units": "fr",
    "value": 1,
  },
  Object {
    "units": "fr",
    "value": 1,
  },
  Object {
    "units": "scalar",
    "value": 20,
  },
  Object {
    "units": "percentage",
    "value": 0.305,
  },
]
`;

exports[`should parse tracks 3`] = `
Array [
  Object {
    "units": "scalar",
    "value": 1,
  },
  Object {
    "units": "percentage",
    "value": 0.02,
  },
  Object {
    "units": "fr",
    "value": 3,
  },
  Object {
    "units": "fr",
    "value": 1,
  },
]
`;
